content/posts/Docker Tags.md

```diff
@@ -0,0 +1,54 @@
+
+So perhaps you decide to use `FROM python:3` and that way you know that your application will always use the latest patched version of python 3. Whenever 3.9, 3.10, 3.11, etc. are released, just by deploying new code you’ll automatically be upgrading your python version along the way. Neat! The downside to this approach is you will likely run into some unexpected failed builds. One day, some package you are using is going to change versions, or have some incompatibility with some other package that changed, and your application will fail to build. This is a frustrating experience to have builds fail when changing code that doesn't touch any package versions.
+
+So maybe you were using `FROM python:3` and one day your builds started failing when the version jumped from `3.7` to `3.8`. You then decide to switch to using `FROM python:3.7` instead. Now, when 3.8, 3.9, 3.10 etc. are released, you won’t automatically bump to those minor release versions without you changing it in your Dockerfile.
+
+You can begin to see where this pattern goes. By starting out at the least specific version tag, you are basically automatically going to be building the newest and latest version of that image. Moving down the spectrum, you can specify a major release versions to use, or the major + minor release version, or even the major + minor + security patch level to use. You also have one additional option: specifying the exact `sha256` fingerprint of an image.
+
+Instead of numbered version tag, you could specify `FROM python@sha256:7e6c00cc553fdce06c1bcfcbf34c73a0f3623a8fc9ce88c8fb91671f2cbc0dab` in your Dockerfile. This is the exact fingerprint of a specific image, and it will never change. This would ensure that nothing about your base image would ever change out from underneath you. This means you probably won't see as many unexpected build failures when some downstream packages. The tradeoff is you won’t be pulling in any newer versions of base images or packages that contain updates, like security patches.
+
+To sum is up: When using something like the `:latest` tag of a public Docker image you are basically always going to be running a patched version of that container, but the packages inside will constantly be changing. By using a specific `sha256` the image won’t change, meaning you won’t get any patches or updates without updating the tag in your Dockerfile.
+
+So which strategy should you use? Ultimately this should be a discussion you will want to have with anyone working on the codebase. Depending on a number of factors such as the number of people working on an application, the resources the team has available, the complexity of the application, and the language used, one tagging strategy may be more appropriate over another.
```

Comment:
Can you very quickly summarise the main points here? Like 'Movable Tag Strategy: Pros: you automatically track upstream patches and updates. Cons: unexpected build breakage. Fixed Digest Strategy: Pros: reproducible builds, no surprises. Cons: it's easy to get a long way behind origin, and miss out on important patches."